### PR TITLE
Pull request for module: lofreq/callparallel

### PR DESCRIPTION
**Pull request by kaurravneet4123**
- [ ] Remove all TODO statements.
- [ ] Emit the `<SOFTWARE>.version.txt` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: software/lofreq/callparallel/main.nf [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        container "https://depot.galaxyproject.org/singularity/lofreq:2.1.5--py38h588ecb2_4"
+    } else {
+        container "quay.io/biocontainers/lofreq:2.1.5--py38h588ecb2_4"
+    }
+
+    input:
+    tuple val(meta), path(bam)
+    file bam_index
+    file fasta
+    file fasta_index
```
Comment (drpatelh) [inline software/lofreq/callparallel/main.nf:25]:
```suggestion
    tuple val(meta), path(bam), path(bai)
    file  fasta
    file  fai
```

File: software/lofreq/callparallel/main.nf [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    lofreq call-parallel \\
+        --pp-threads ${task.cpus} \\
+        -f $fasta \\
+        -o ${prefix}.vcf \\
+        $bam
```
Comment (drpatelh) [inline software/lofreq/callparallel/main.nf:39]:
```suggestion
    lofreq \\
        call-parallel \\
        --pp-threads $task.cpus \\
        -f $fasta \\
        -o ${prefix}.vcf \\
        $bam
```

File: software/lofreq/callparallel/meta.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test' ]
+  - bam:
+      type: file
+      description: Sorted BAM file
+      pattern: "*.{bam}"
+  - bam_index:
```
Comment (drpatelh) [inline software/lofreq/callparallel/meta.yml:24]:
```suggestion
  - bai:
```

File: software/lofreq/callparallel/meta.yml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      type: file
+      description: BAM index file
+      pattern: "*.{bai}"
+  - fasta:
+      type: file
+      description: Reference genome FASTA file
+      pattern: "*.{fasta}"
+  - fasta_index:
```
Comment (drpatelh) [inline software/lofreq/callparallel/meta.yml:32]:
```suggestion
  - fai:
```

File: tests/software/lofreq/callparallel/main.nf [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    
+    input = [ [ id:'test' ], // meta map
+              file(params.test_data['sarscov2']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true) ]
+    
+    fasta       = file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
+    
+    SAMTOOLS_INDEX ( input )
+    SAMTOOLS_FAIDX ( fasta )
+
+    LOFREQ_CALLPARALLEL ( input, SAMTOOLS_INDEX.out.bai, fasta, SAMTOOLS_FAIDX.out.fai )
```
Comment (drpatelh) [inline tests/software/lofreq/callparallel/main.nf:19]:
```suggestion
include { LOFREQ_CALLPARALLEL } from '../../../../software/lofreq/callparallel/main.nf' addParams( options: [:] )

workflow test_lofreq_callparallel {
    
    input = [ 
        [ id:'test' ], // meta map
        file(params.test_data['sarscov2']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true),
        file(params.test_data['sarscov2']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExists: true)
    ]
    
    fasta = file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
	fai   = file(params.test_data['sarscov2']['genome']['genome_fasta_fai'], checkIfExists: true)
    
    LOFREQ_CALLPARALLEL ( input, fasta, fai )
```

**Comment by kaurravneet4123** [comment]
Thanks for super quick feedback @drpatelh :-)

File: software/lofreq/callparallel/main.nf (hunk 1 above)
Comment (drpatelh) [reply software/lofreq/callparallel/main.nf:25]: The `bam` and `bai` are normally shipped together for each sample so should be defined in the same channel. Does `lofreq` fail without the BAM index and FASTA index?

Comment (kaurravneet4123) [reply software/lofreq/callparallel/main.nf]: Yes, lofreq fails without the BAM index and FASTA index. I have now defined the bam and bai in the same channel and it is working fine.

File: software/lofreq/callparallel/main.nf (hunk 2 above)
Comment (kaurravneet4123) [reply software/lofreq/callparallel/main.nf]: changes addressed

File: software/lofreq/callparallel/meta.yml (hunk 3 above)
Comment (kaurravneet4123) [reply software/lofreq/callparallel/meta.yml]: change done!

File: tests/software/lofreq/callparallel/main.nf (hunk 5 above)
Comment (kaurravneet4123) [reply tests/software/lofreq/callparallel/main.nf]: change done!

File: tests/software/lofreq/callparallel/test.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+- name: lofreq callparallel
+  command: nextflow run ./tests/software/lofreq/callparallel -entry test_lofreq_callparallel -c tests/config/nextflow.config
+  tags:
+    - lofreq
+    - lofreq/callparallel
+  files:
+    - path: output/lofreq/test.vcf
```
Comment (drpatelh) [inline tests/software/lofreq/callparallel/test.yml:7]: You removed the md5sum here?

File: software/lofreq/callparallel/main.nf [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    tuple val(meta), path("*.vcf"), emit: vcf
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    lofreq \\
+        call-parallel \\
+        --pp-threads ${task.cpus} \\
```
Comment (drpatelh) [inline software/lofreq/callparallel/main.nf:36]:
```suggestion
        --pp-threads $task.cpus \\
```

File: tests/software/lofreq/callparallel/test.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+      md5sum: 5bfa1c927ab131c6dc1109bb41dc5992
```
Comment (drpatelh) [inline tests/software/lofreq/callparallel/test.yml:8]:
```suggestion
```

Comment (drpatelh) [reply tests/software/lofreq/callparallel/test.yml:8]: Ok. So it appears that there is something variable that is added to the VCF file which is causing the md5sum checks to fail.

**Review by drpatelh — APPROVED** [review]

**Comment by kaurravneet4123** [comment]
Hi @drpatelh. Apologies as I could not reply earlier as it has been a very busy day at work! Many thanks for your feedback and for merging the module. Regards, Rav

**Pull request merged**